Convert unit

yunzhu1229 opened the pull request:
Hi Fede,
I have added the "switch-si-ip" function in the new commit, the "convert_data" function can be called along with this function. The mapping_dictionary will also get changed in the convert function.
All the data is stored in the local and it only need to do the conversion when the "si-ip" button is clicked. 
Based on the "switch-si-ip" function, I have modified the first seven pages.
Please help me review it.
Thanks

yunzhu1229 commented:
Hi Fede,
I have added the 'si' and 'ip' attributes in the mapping_dictionary and removed the mapping_dictionary_store. So that we can access the units by using mapping_dictionary[var][si_ip][“unit’] as you suggested. All the contents of the website also have been adjusted; they can work correctly when switching the units. Please help me review and confirm it. Thanks a lot.

FedericoTartarini commented:
Great work, I will try to review it as soon as possible and get back to you if I have any comments

FedericoTartarini commented on my_project/extract_df.py:
```diff
@@ -306,6 +308,49 @@ def create_df(lst, file_name):
 
+#convert function
+def temperature(df,name):
+    df[name] = df[name]*1.8+32
```
@yunzhu1229 if I understand correctly this function can only convert from C to F but cannot go back. Is this the case? What if the user wants to go back to C? Maybe it is not important to implement this feature but you should discuss this with @stefanoschiavon.

yunzhu1229 replied:
In the switch_si_ip function, I have defined the data request method. When the state go back to 'si', it will directly request the data again by the url, so the data will go back to 'si' state and needn't do the conversion. So I think the user can flexibly choose the state, all the data can be converted from 'si' to 'ip', and they can also go back to 'si'.

FedericoTartarini replied:
okay, thank you for the explanation. The issue with this approach is that we are using resources inefficiently, this will generate an unnecessary network request to download the data and also it will slow down the application. We should either save the SI unit dataframe locally so it can be recalled if the user goes back to the SI selection or implement functions to convert the data back to SI.

yunzhu1229 replied:
See the other comments. Because the df-store can only be output once, so I think it should be output at the switch_si_ip function, so that the data after conversion can be used elsewhere. Therefore, it is needed to create the df in the switch_si_ip function. At the same time, the back convert from PI to SI is no needed, it can be done just by create the df again.

FedericoTartarini commented on my_project/global_scheme.py:
```diff
@@ -133,20 +132,41 @@
+            "unit": "°F",
+            "range": [-40, 122],
+        },
+        "conversion_function": "temperature",
```
This is very good since now you can select the unit to display and the range as a function of the user's selection

FedericoTartarini commented on my_project/tab_natural_ventilation/app_natural_ventilation.py:
```diff
@@ -85,7 +86,7 @@ def inputs_tab():
                         n_clicks=1,
                     ),
-                    html.H6("Outdoor dry-bulb air temperature range"),
+                    html.H6("Outdoor dry-bulb air temperature range -- (default: °C) / °F"),
```
this should not be hardcoded but it should be updated as a function of the user selection

yunzhu1229 replied:
This problem is now solved. The set number will be changed automatically according to the si-ip-state.

FedericoTartarini commented on my_project/tab_natural_ventilation/app_natural_ventilation.py:
```diff
@@ -202,7 +203,7 @@ def inputs_tab():
                         options=[
                             {
-                                "label": "Avoid condensation with radiant systems: If the outdoor dew point temperature is below the radiant system surface temperature, the data point is not plot.",
+                                "label": "Avoid condensation with radiant systems: If the outdoor dew point temperature is below the radiant system surface temperature, the data point is not plot. -- (default: °C) / °F",
```
same as the previous comment. This should not be hardcoded.

yunzhu1229 replied:
This problem is now solved. The set number will be changed automatically according to the si-ip-state.

FedericoTartarini commented on my_project/tab_psy_chart/app_psy_chart.py:
```diff
@@ -215,6 +216,7 @@ def layout_psy_chart():
 @app.callback(
     Output("psych-chart", "children"),
     [
+        Input("df-store", "modified_timestamp"),
```
Are you using this ts information?

yunzhu1229 replied:
This is needed. Because when the user changes the si-ip state on this page, all the data and graphs need to change, so ts information is used to listen for the changes.

FedericoTartarini commented on my_project/tab_select/app_select.py:
```diff
@@ -127,15 +132,13 @@ def submitted_data(
         lines = get_data(url_store)
```
@yunzhu1229 I identified another problem, here, you are requesting the data and then you are requesting again in the switch_ip_si function. This will slow down the code significantly.

yunzhu1229 replied:
I have modified the function switch_si_ip. If the state is 'ip', it only does the conversion and will not request the data. If it is 'si', it will request the data. In this way, it can also solve the problem of changing data back from 'ip' state to 'si' state.

FedericoTartarini replied:
see other comment, about this approach

yunzhu1229 replied:
I understand your meanings, and I have tried to fix the problem. However, I found it is necessary to create df in the switch_si_ip function, because it is needed to output the df-store and the df-store can only be output once in the program. To avoid requesting the data twice, I have stored the lines as a store, it can save the running time to some extent.

FedericoTartarini replied:
okay great, @yunzhu1229 have you pushed the changes to the brach so @danielh7-cs9 and his team can start work on it?

FedericoTartarini commented on my_project/tab_select/app_select.py:
```diff
@@ -172,14 +173,45 @@ def submitted_data(
+
+def switch_si_ip(si_ip_input, meta, url_store):
+    si_ip_store = si_ip_input
+    lines = get_data(url_store)
```
here you are requesting the data again and I guess this is slowing down the code significantly

yunzhu1229 replied:
See other comments.

FedericoTartarini commented on my_project/tab_summary/app_summary.py:
```diff
@@ -89,7 +91,7 @@ def layout_summary():
                             dbc.Col(
                                 html.Label(
-                                    "Heating degree day (HDD) setpoint (°C)",
+                                    "Heating degree day (HDD) setpoint (default: °C) / °F",
```
This should be updated as a function of the user selection

FedericoTartarini replied:
As you have done below in line 224

yunzhu1229 replied:
This problem is now solved. The set number will be changed automatically according to the si-ip-state.

FedericoTartarini replied:
please push the changes since you are mentioning that the problem is solved but I cannot see the new code.

FedericoTartarini commented on my_project/tab_sun/app_sun.py:
```diff
@@ -131,17 +132,17 @@ def static_section():
-                type="circle",
-                children=html.Div(id="monthly-solar"),
-            ),
+            # html.Div(
```
Why you have commented this code out?

FedericoTartarini replied:
I am just reviewing the code from GitHub so I cannot verify this for you but I am going to close this issue if you are saying that the code is duplicated

FedericoTartarini commented on my_project/tab_sun/app_sun.py:
```diff
@@ -195,10 +196,11 @@ def monthly_and_cloud_chart(ts, df, meta):
+    #     figure=monthly,
+    # ), 
     return dcc.Graph(
-        config=generate_chart_name("monthly_sun", meta),
```
why have you removed this code?

FedericoTartarini commented on my_project/template_graphs.py:
```diff
@@ -394,8 +399,16 @@ def wind_rose(df, title, month, hour, labels):
     spd_colors = mapping_dictionary["wind_speed"]["color"]
+    spd_unit = mapping_dictionary["wind_speed"][si_ip]["unit"]
     spd_bins = [-1, 0.5, 1.5, 3.3, 5.5, 7.9, 10.7, 13.8, 17.1, 20.7, np.inf]
-    spd_labels = speed_labels(spd_bins, units="m/s")
```
this conversion should not happen here nested inside a chart

yunzhu1229 replied:
I have added a conversion function for the spd_bins.

FedericoTartarini reviewed (commented):
Please address my comments

yunzhu1229 commented on my_project/extract_df.py:
```diff
@@ -306,6 +308,76 @@ def create_df(lst, file_name):
+    df["adaptive_cmf_90_low"] = df["adaptive_cmf_90_low"]*1.8+32
+    df["adaptive_cmf_90_up"] = df["adaptive_cmf_90_up"]*1.8+32
+
+    mapping_dict = json.loads(mapping_json)
```
Currently, the mapping dictionary is used as a constant. I didn't make changes to it in other functions.

yunzhu1229 commented on my_project/tab_data_explorer/app_data_explorer.py:
```diff
@@ -549,7 +549,7 @@ def layout_data_explorer():
     [State("df-store", "data"), State("meta-store", "data")],
 )
-def update_tab_yearly(var, global_local, df, meta):
+def update_tab_yearly(var, global_local, si_ip, df, meta):
```
This version of the code seems outdated. In the newest commit, I have added the 'si-ip-unit-store' as a State.
-----
   [State("df-store", "data"), State("meta-store", "data"), State("si-ip-unit-store", "data")],

def update_tab_yearly(ts, var, global_local, df, meta, si_ip):

yunzhu1229 commented on my_project/tab_data_explorer/app_data_explorer.py:
```diff
@@ -623,6 +623,7 @@ def update_tab_heatmap(var, global_local, df, meta):
         State("meta-store", "data"),
         State("invert-month-explore-heatmap", "value"),
         State("invert-hour-explore-heatmap", "value"),
+        State("map-dictionary-store","data"),
```
This version of the code seems outdated. In the newest version, the map-dictionary-store is removed.

yunzhu1229 commented on my_project/tab_sun/charts_sun.py:
```diff
@@ -96,18 +98,19 @@ def monthly_solar(epw_df):
-        var_range = mapping_dictionary[var]["range"]
-        var_name = mapping_dictionary[var]["name"]
-        var_color = mapping_dictionary[var]["color"]
+        var_unit = map_dict[var]["unit"]
```
This code is outdated. I have modified the mapping_dictionary, using the logic map_dict[var][si_ip][unit]

yunzhu1229 commented on my_project/tab_sun/app_sun.py:
```diff
@@ -168,24 +168,24 @@ def layout_sun():
         Input("df-store", "modified_timestamp"),
     ],
-    [State("df-store", "data"), State("meta-store", "data")],
+    [State("df-store", "data"), State("meta-store", "data"), State("map-dictionary-store","data")],
```
This code is outdated. I have removed the map-dictionary-store.

FedericoTartarini commented:
@yunzhu1229 thank you again for working on the code. I have a team of students who are going to be working on the project and helping us to fix all the issues. Could you please then push to GitHub all the changes you have made? I will then merge your changes into the main repository and ask them to review your code. In this way, they can also start working on other issues and you can all collaborate on the code. I am asking for this since you edited a lot of files and I am afraid that if they start working on the source code then we will face a lot of merging issues when we merge your pull request. Could you please push your code changes my this Sunday and let me know once that is done?

yunzhu1229 commented:
Hi Fede, Thank you for arranging this. I will pull request for the current
work.

Currently, I have addressed all the comments you replied. I have also
changed the data request way and added the function that can turn the ‘ip’
state back into ‘si’ state. It works. However, after doing this some graph
will have some problems, because some graph are editing in a specific way,
some are using the specific number for drawing, that will cause some
problems when the data are converted by unit system. I need to look through
the code and modify them. It may take some time.

By the way, I will check about the results with Stefano and Hui next week.
We have plan a meeting at March 15th, because they also mentioned some
problems. If I pull request the work, and others continued work on that,
did it mean I can’t modify them if I need to do some more modifications.

On Thu, Mar 9, 2023 at 18:39 Federico Tartarini ***@***.***>
wrote:

> @yunzhu1229 <https://github.com/yunzhu1229> thank you again for working
> on the code. I have a team of students who are going to be working on the
> project and helping us to fix all the issues. Could you please then push to
> GitHub all the changes you have made? I will then merge your changes into
> the main repository and ask them to review your code. In this way, they can
> also start working on other issues and you can all collaborate on the code.
> I am asking for this since you edited a lot of files and I am afraid that
> if they start working on the source code then we will face a lot of merging
> issues when we merge your pull request. Could you please push your code
> changes my this Sunday and let me know once that is done?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/CenterForTheBuiltEnvironment/clima/pull/146#issuecomment-1463151428>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYHJ7R4FUJPSAFVCHYIMBMTW3KH6VANCNFSM6AAAAAAUAA4ONA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

FedericoTartarini commented:
Thank you so much and thank you for fixing all the bugs you mentioned. 

Please just push the changes that you have made locally to GitHub so the branch contains your latest edits. The other students here can also help you troubleshoot the issues you are currently facing. Once all the changes are going to be integrated we can then merge the pull request. 

I have tagged @danielh7-cs9, he is leading the other students and they will be able to help you, so do not worry if you cannot fix everything. I will let you @yunzhu1229 and @danielh7-cs9 discuss how to move forward. Daniel may also have some futher questions for you.

yunzhu1229 commented:
Dear all,
I have committed the newest codes in GitHub for the new team member to work
on that.
I have fixed the problems you mentioned before. I also fixed some other
problems with the previous graph-making work so that the contents can
change flexibly with the unit state.
Thanks to Federico, he helps me review the work and make comments on the
code. I have checked all the comments and made modifications accordingly.
I will show the result at next week's meeting. Thanks again for all your
help.

Regards,
Yunzhu

yunzhu1229 commented:
Dear Yunzhu,
Thank you for the great work on this!  Looking forward to seeing it at the
Thursday meeting!
Hui

On Sun, Mar 12, 2023 at 8:01 PM Yunzhu Ji ***@***.***> wrote:

> Dear all,
> I have committed the newest codes in GitHub for the new team member to
> work on that.
> I have fixed the problems you mentioned before. I also fixed some other
> problems with the previous graph-making work so that the contents can
> change flexibly with the unit state.
> Thanks to Federico, he helps me review the work and make comments on the
> code. I have checked all the comments and made modifications accordingly.
> I will show the result at next week's meeting. Thanks again for all your
> help.
>
> Regards,
> Yunzhu
>
>

danielh7-cs9 commented:
@FedericoTartarini Just to confirm requirements for the CS9 team.
1. Team is to pull down the the convert-unit code into local IDE.
2. Implement UAT and report any issues based on the new code? If identified, log issue and investigate further.

Thanks,
Dan

FedericoTartarini commented:
@danielh7-cs9 correct it would be great if the team could:
* pull the code and test it locally
* make sure there are no bugs and that a hypothetical user can easily toggle between the SI and IP units.
* Review and suggest if there are any potential solutions to improve the code base.
* check if the responsiveness of the new version of the app (new beach) is the same as the original one (main beach)
Please let me know if you have any questions.

danielh7-cs9 commented:
> @danielh7-cs9 correct it would be great if the team could:
> 
> * pull the code and test it locally
> * make sure there are no bugs and that a hypothetical user can easily toggle between the SI and IP units.
> * Review and suggest if there are any potential solutions to improve the code base.
> * check if the responsiveness of the new version of the app (new beach) is the same as the original one (main beach)
> 
> Please let me know if you have any questions. Once all the above are done, we can create a new pull request and deploy a new version of Clima

Thank you. Will coordinate this work with the team.

phuongdoan13 commented:
I can easily changed between SI and IP

phuongdoan13 commented:
Not sure if this one has been brought up to you. I noticed that the code contained repetition of raw string such as df['hour'], df['minute']. These repetitions can cause run-time errors(i.e. errors that are not caught during "compile" time, but only when the user interact with the UI) if someone accidentally mispell a string (e.g. "hours" instead of "hour"). 

A solution I learned at the workplace is putting these raw strings into variables. hour = "hour", minute ="minute". Putting these variable in classes if necessary. This will avoid repeting raw string, as well as allowing IDE suggestions

yunzhu1229 commented:
Hi Stefano!

Yes, Yunzhu and I are already arranging a meeting for that, which will most
likely take place this Tuesday.

Marcello




*Mag. Arch. Marcello Turrini ZT.*
*----------------------------------------------------------------------------------------------*
*Sustainable Architecture and Urban Planning*
Schlachthausgasse 38b/34
1030 Wien
AUSTRIA

*PhD Candidate*
*On Climate Change Resilience and Design of Nature Based Solutions*
Università di Parma - Dipartimento di Ingegneria e Architettura - DIA
Parco Area delle Scienze 181/a
43124 Parma
ITALY






Il giorno lun 13 mar 2023 alle ore 09:05 Stefano Schiavon <
***@***.***> ha scritto:

> Hi Yunzhu, thank you!
> Hi Marcello, could you carefully check the tool and let us know if you
> find any errors?
>
> Stefano
>
>
> On Sun, Mar 12, 2023 at 8:01 PM Yunzhu Ji ***@***.***> wrote:
>
>> Dear all,
>> I have committed the newest codes in GitHub for the new team member to
>> work on that.
>> I have fixed the problems you mentioned before. I also fixed some other
>> problems with the previous graph-making work so that the contents can
>> change flexibly with the unit state.
>> Thanks to Federico, he helps me review the work and make comments on the
>> code. I have checked all the comments and made modifications accordingly.
>> I will show the result at next week's meeting. Thanks again for all your
>> help.
>>
>> Regards,
>> Yunzhu
>>
>>

yunzhu1229 commented:
Great, please then list all the issues you will identify either in the Pull request or please open separate issues on GitHub, so the team of students from the University of Sydney can work on them.

From: Turrini Marcello ***@***.***>
Date: Tuesday, 14 March 2023 at 3:23 am
To: Stefano Schiavon ***@***.***>
Cc: Yunzhu Ji ***@***.***>, CenterForTheBuiltEnvironment/clima ***@***.***>, Federico Tartarini ***@***.***>, Hui Zhang ***@***.***>, David R. Lehrer ***@***.***>
Subject: Re: Pull request for the modification
Hi Stefano!

Yes, Yunzhu and I are already arranging a meeting for that, which will most likely take place this Tuesday.

Marcello




Mag. Arch. Marcello Turrini ZT.
----------------------------------------------------------------------------------------------
Sustainable Architecture and Urban Planning
Schlachthausgasse 38b/34
1030 Wien
AUSTRIA

PhD Candidate
On Climate Change Resilience and Design of Nature Based Solutions
Università di Parma - Dipartimento di Ingegneria e Architettura - DIA
Parco Area delle Scienze 181/a
43124 Parma
ITALY





Il giorno lun 13 mar 2023 alle ore 09:05 Stefano Schiavon ***@***.******@***.***>> ha scritto:
Hi Yunzhu, thank you!
Hi Marcello, could you carefully check the tool and let us know if you find any errors?

Stefano


On Sun, Mar 12, 2023 at 8:01 PM Yunzhu Ji ***@***.******@***.***>> wrote:
Dear all,
I have committed the newest codes in GitHub for the new team member to work on that.
I have fixed the problems you mentioned before. I also fixed some other problems with the previous graph-making work so that the contents can change flexibly with the unit state.
Thanks to Federico, he helps me review the work and make comments on the code. I have checked all the comments and made modifications accordingly.
I will show the result at next week's meeting. Thanks again for all your help.

Regards,
Yunzhu

FedericoTartarini commented:
> Not sure if this one has been brought up to you. I noticed that the code contained repetition of raw string such as df['hour'], df['minute']. These repetitions can cause run-time errors(i.e. errors that are not caught during "compile" time, but only when the user interact with the UI) if someone accidentally mispell a string (e.g. "hours" instead of "hour").
> 
> A solution I learned at the workplace is putting these raw strings into variables. hour = "hour", minute ="minute". Putting these variable in classes if necessary. This will avoid repeting raw string, as well as allowing IDE suggestions

I really like the solution of putting these string variables in classes. Please feel free to edit the code.

yunzhu1229 commented on my_project/tab_psy_chart/app_psy_chart.py:
```diff
@@ -308,26 +313,25 @@ def update_psych_chart(
-        var_range_y = hr_range
+        var_range_x = mapping_dictionary["DBT"][si_ip]["range"]
+        var_range_y = mapping_dictionary["hr"][si_ip]["range"]
 
```
I changed the range according to the mapping_dictionary, so it can change when the unit system changes.

yunzhu1229 commented on my_project/tab_psy_chart/app_psy_chart.py:
```diff
@@ -308,26 +313,25 @@ def update_psych_chart(
-        data_max = (5 * ceil(df["hr"].max() * 1000 / 5)) / 1000
-        data_min = (5 * floor(df["hr"].min() * 1000 / 5)) / 1000
+        data_max = round(df["hr"].max(),4)
+        data_min = round(df["hr"].min(),4)
```
I changed the max and min data acquisition methods. The previous method may lead to some deviation when the unit system changes.

yunzhu1229 commented on my_project/tab_psy_chart/app_psy_chart.py:
```diff
@@ -350,10 +354,19 @@ def update_psych_chart(
+                dbt_list_convert[j] = dbt_list_convert[j]*1.8+32
+            for k in range(len(rh_df[name])):
+                rh_convert[k] = rh_convert[k]*0.0624
         fig.add_trace(
```
Since the data here are extracted from the 'pythermalcomfort' (another data source), not the 'df-store'; so I added the additional data conversion method here.

yunzhu1229 commented on my_project/tab_sun/charts_sun.py:
```diff
@@ -130,7 +131,7 @@ def polar_graph(df, meta, global_local, var):
     else:
         vals = solpos[var]
-        marker_size = (((vals - vals.min()) / vals.max()) + 1) * 4
+        marker_size = ((vals - vals.min()) / (vals.max()-vals.min()) + 1) * 4
```
I changed the definition for the 'maker size'; so that can avoid the too-big maker when the unit system is changed to 'IP'.

yunzhu1229 commented on my_project/template_graphs.py:
```diff
@@ -157,7 +161,7 @@ def yearly_profile(df, var, global_local):
             hovertemplate=(
-                "Max: %{y:.2f} &#8451;<br>" + "Min: %{base:.2f} &#8451;<br>"
+                "Max: %{y:.2f} " + var_unit + "Min: %{base:.2f} " + var_unit
             ),
```
The unit for 'adaptive comfort 80%' and 'adaptive comfort 90%' is defined statically, and I changed it to refer to the mapping_dictionary.

yunzhu1229 commented:
Hi @danielh7-cs9 ,
I have added a new commit for bringing back the sun radiation graph. Please check it. Thank you.

danielh7-cs9 commented:
> @danielh7-cs9 correct it would be great if the team could:
> 
> * pull the code and test it locally
> * make sure there are no bugs and that a hypothetical user can easily toggle between the SI and IP units.
> * Review and suggest if there are any potential solutions to improve the code base.
> * check if the responsiveness of the new version of the app (new beach) is the same as the original one (main beach)
> 
> Please let me know if you have any questions. Once all the above are done, we can create a new pull request and deploy a new version of Clima

Hi @FedericoTartarini , the team have spent time conducting UAT on yunzhu's code and we've currently not identified any bugs or anomalies with performance so far. Suggest that once all code is merged if there are issues we log it as an issue and look to add into our project plan.

FedericoTartarini commented:
Great to hear that, I will then merge the Pull request into the main branch

danielh7-cs9 commented:
> > Not sure if this one has been brought up to you. I noticed that the code contained repetition of raw string such as df['hour'], df['minute']. These repetitions can cause run-time errors(i.e. errors that are not caught during "compile" time, but only when the user interact with the UI) if someone accidentally mispell a string (e.g. "hours" instead of "hour").
> > A solution I learned at the workplace is putting these raw strings into variables. hour = "hour", minute ="minute". Putting these variable in classes if necessary. This will avoid repeting raw string, as well as allowing IDE suggestions
> 
> I really like the solution of putting these string variables in classes. Please feel free to edit the code.

@phuongdoan13 - Let's log this as an issue (enhancement) so we can track it appropriately.

yunzhu1229 commented:
Dear all,

Today we met with Stefano, Hui, David, and Marcello. We checked all the
contents of the Clima Tool. Currently, the unit conversion function is
basically finished, and we also found some problems.

I have quickly modified some of them, and some need Stefano to confirm.
Others may need further modification. I have sorted out all the problems in
the attached file.

I hope all this information will be useful for your following work, and I
will also pull a request for my last changed code on GitHub and comment on
it.

Thank you for everyone's support for my work.


Regards,

Yunzhu

yunzhu1229 commented:
Hi Federico and team,

I reviewed the suggestions from Yunzhu and saved in this shared google file
<https://docs.google.com/document/d/1SHctEw6Rq48iAGx6WicXl77YcM3Yf08R/edit?usp=sharing&ouid=117385117068999316356&rtpof=true&sd=true>
(Yunzhu, it is easier to collaborate on a shared online file instead of
sending a file as an attachment to the email).

Best,
Stefano



On Fri, Mar 17, 2023 at 12:52 AM Yunzhu Ji ***@***.***> wrote:

> Dear all,
>
> Today we met with Stefano, Hui, David, and Marcello. We checked all the
> contents of the Clima Tool. Currently, the unit conversion function is
> basically finished, and we also found some problems.
>
> I have quickly modified some of them, and some need Stefano to confirm.
> Others may need further modification. I have sorted out all the problems in
> the attached file.
>
> I hope all this information will be useful for your following work, and I
> will also pull a request for my last changed code on GitHub and comment on
> it.
>
> Thank you for everyone's support for my work.
>
>
> Regards,
>
> Yunzhu
>
